File: src/main/java/com/ecommerce/couponservice/redis/scheduler/CouponIssuanceScheduler.java
package com.ecommerce.couponservice.redis.scheduler;

import com.ecommerce.couponservice.domain.coupon.repo.CouponRedisRepository;
import com.ecommerce.couponservice.domain.coupon.service.CouponIssuanceService;
import jakarta.persistence.EntityNotFoundException;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.redis.core.ZSetOperations;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Objects;
import java.util.Set;

@Service
@RequiredArgsConstructor
@Slf4j
public class CouponIssuanceScheduler extends BaseCouponScheduler {

    private final CouponRedisRepository couponRedisRepository;
    private final CouponIssuanceService couponIssuanceService;

    @Scheduled(fixedDelay = 5_000, initialDelay = 13_000)
    private void processScheduleEnterQueueTasks() {
        Set<String> enterKeys = couponRedisRepository.getEnterQueueKeys();
        if(enterKeys.isEmpty()) {
            log.debug("No enter queues found to process.");
            return;
        }

        for(String enterKey: enterKeys) {
            processEnterQueue(enterKey);
        }
    }

    private void processEnterQueue(String enterKey) {
        try {
            Long couponId = extractCouponId(CouponRedisRepository.ENTER_KEY_PREFIX, enterKey);
            if(couponId == null) {
                log.warn("Invalid enter queue key format: {}", enterKey);
                return;
            }

            List<Long> accountIds = getAccountIdsFromEnterQueue(couponId);
            if(accountIds.isEmpty()) {
                log.warn("No valid accountIds found for couponId: {}", couponId);
                return;
            }
            accountIds.forEach(i -> issueCouponsInBatch(couponId, i));

        } catch (NumberFormatException e) {
            log.error("Failed to parse enter queue key: {}", enterKey, e);
        }  catch (Exception e) {
            log.error("Failed to process enter queue key: {}", enterKey, e);
        }
    }

    List<Long> getAccountIdsFromEnterQueue(Long couponId) {
        return couponRedisRepository.getEnterQueueValueAndScore(couponId)
                .stream()
                .map(ZSetOperations.TypedTuple::getValue)
                .filter(Objects::nonNull)
                .map(Long::parseLong)
                .toList();
    }

    private void issueCouponsInBatch(long couponId, long accountId) {
        try {
            couponIssuanceService.issueCoupon(couponId, accountId);
        } catch (EntityNotFoundException e) {
            log.error("Coupon not found: {}", couponId, e);
        } catch (IllegalArgumentException e) {
            log.error("All coupons({}) have been redeemed.", couponId, e);
        } finally {
            couponRedisRepository.removeEnterQueueValue(couponId, accountId);
        }
    }
}

File: src/main/kotlin/com/ecommerce/couponservice/redis/scheduler/CouponIssuanceScheduler.kt
package com.ecommerce.couponservice.redis.scheduler

import com.ecommerce.couponservice.domain.coupon.service.CouponIssuanceService
import com.ecommerce.couponservice.redis.manager.CouponQueueRedisManager
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Component

@Component
class CouponIssuanceScheduler(
    private val couponQueueRedisManager: CouponQueueRedisManager,
    private val couponIssuanceService: CouponIssuanceService
) : BaseCouponScheduler() {

    @Scheduled(fixedDelay = 5_000, initialDelay = 13_000)
    suspend fun processScheduleEnterQueueTasks() {
        val enterKeys = couponQueueRedisManager.enterQueueKeys
        if (enterKeys.isEmpty()) {
            log.debug("No enter queues found to process.")
            return
        }

        for (enterKey in enterKeys) {
            processEnterQueue(enterKey)
        }
    }

    private suspend fun processEnterQueue(enterKey: String) {
        try {
            val couponId = extractCouponId(CouponQueueRedisManager.ENTER_KEY_PREFIX, enterKey)
            if (couponId == null) {
                log.warn("Invalid enter queue key format: {}", enterKey)
                return
            }

            val accountIds = getAccountIdsFromEnterQueue(couponId)
            if (accountIds.isEmpty()) {
                log.warn("No valid accountIds found for couponId: {}", couponId)
                return
            }
            couponIssuanceService.issueCouponInBatchAsync(couponId, accountIds)
        } catch (e: NumberFormatException) {
            log.error("Failed to parse enter queue key: {}", enterKey, e)
        } catch (e: Exception) {
            log.error("Failed to process enter queue key: {}", enterKey, e)
        }
    }

    private fun getAccountIdsFromEnterQueue(couponId: Long): List<Long> {
        return couponQueueRedisManager.getEnterQueueValueAndScore(couponId)
            .mapNotNull { it.value }
            .mapNotNull { it.toLongOrNull() }
    }

    companion object {
        private val log = org.slf4j.LoggerFactory.getLogger(CouponIssuanceScheduler::class.java)
    }
}